src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/TypeCast.cs

```diff
@@ -42,119 +42,147 @@ internal enum AssignmentVariation
+                    goto done;
+
+                if (mt == null)
+                    break;
+
+                mt = mt->NonArrayBaseType;
+                if (mt == pTargetType)
+                    goto done;
+
+                if (mt == null)
+                    break;
+
+                mt = mt->NonArrayBaseType;
+                if (mt == pTargetType)
+                    goto done;
+
+                if (mt == null)
+                    break;
+
+                mt = mt->NonArrayBaseType;
             }
 
-            return IsInstanceOfClass_Helper(pTargetType, obj, pObjType);
+        fail:
+            obj = null;
+
+        done:
+            return obj;
         }
 
-        private static unsafe object IsInstanceOfClass_Helper(MethodTable* pTargetType, object obj, MethodTable* pObjType)
+        [RuntimeExport("RhTypeCast_CheckCastClass")]
+        public static unsafe object CheckCastClass(MethodTable* pTargetType, object obj)
         {
-            // if the EETypes pointers match, we're done
-            if (pObjType == pTargetType)
+            Debug.Assert(!pTargetType->IsParameterizedType, "CheckCastClass called with parameterized MethodTable");
+            Debug.Assert(!pTargetType->IsFunctionPointerType, "CheckCastClass called with function pointer MethodTable");
+            Debug.Assert(!pTargetType->IsInterface, "CheckCastClass called with interface MethodTable");
+            Debug.Assert(!pTargetType->HasGenericVariance, "CheckCastClass with variant MethodTable");
+
+            if (obj == null || obj.GetMethodTable() == pTargetType)
             {
                 return obj;
             }
 
-            if (pTargetType->HasGenericVariance && pObjType->HasGenericVariance)
-            {
-                // Only generic interfaces and delegates can have generic variance and we shouldn't see
-                // interfaces for either input here. So if the canonical types are marked as having variance
-                // we know we've hit the delegate case. We've dealt with the identical case just above. And
-                // the regular path below will handle casting to Object, Delegate and MulticastDelegate. Since
-                // we don't support deriving from user delegate classes any further all we have to check here
-                // is that the uninstantiated generic delegate definitions are the same and the type
-                // parameters are compatible.
+            return CheckCastClassSpecial(pTargetType, obj);
+        }
 
-                // NOTE: using general assignable path for the cache because of the cost of the variance checks
-                if (AreTypesAssignableInternal(pObjType, pTargetType, AssignmentVariation.BoxedSource, null))
-                    return obj;
-                return null;
-            }
+        [RuntimeExport("RhTypeCast_CheckCastClassSpecial")]
+        private static unsafe object CheckCastClassSpecial(MethodTable* pTargetType, object obj)
```

Comment:
It would be nice to unify the names with CoreCLR impl (Check vs. Chk) so that the implementations are easy to match (and eventually share). It can be separate rename-only PR.